### PR TITLE
[WIP] Move rule generation under fiber

### PR DESCRIPTION
**Pull request by rgrinberg**
- [ ] Traverse file_tree in the monad
- [ ] Expand inside the monad (Or not. There's still some discussion here)

### CODE REVIEW

**Comment by diml** [comment]
I was thinking about this earlier. Regarding the expander, I was thinking that it'd be nice to introduce a monad `Fiber_and_deps.t` that would allow to collect dependencies statically. I.e. with primitives like this:

```ocaml
val static_deps : Dep.Set.t -> unit t

val get_static_deps : _ t -> Dep.Set.t
```

This way, we wouldn't need the "partial" functions. That would simplify the code.

**Comment by diml** [comment]
An applicative actually, a monad wouldn't work here

**Comment by aalekseyev** [comment]
That sounds very similar to what `Build.t` is currently, since the targets specification was already moved into a separate type (`Build.With_targets.t`). That said, I'm not sure what problem you're referring to (what partial functions).

File: src/dune/build_system.mli [hunk 1]
```diff
@@ -79,13 +81,13 @@ val prefix_rules : unit Build.t -> f:(unit -> 'a) -> 'a
...
-val eval_pred : File_selector.t -> Path.Set.t
+val eval_pred : File_selector.t -> Path.Set.t Fiber.t
 
 (** Returns the set of targets in the given directory. *)
-val targets_of : dir:Path.t -> Path.Set.t
+val targets_of : dir:Path.t -> Path.Set.t Fiber.t
 
 (** Load the rules for this directory. *)
-val load_dir : dir:Path.t -> unit
+val load_dir : dir:Path.t -> unit Fiber.t
```
Comment (rgrinberg) [inline src/dune/build_system.mli]: @diml have a look at the changes in this mli. As far as I can tell, the callback to gen_rules can call all the functions above so we are required to change the return type of the callback to `unit Fiber.t`.

**Comment by diml** [comment]
> That sounds very similar to what Build.t is currently, since the targets specification was already moved into a separate type (Build.With_targets.t). That said, I'm not sure what problem you're referring to (what partial functions).

Indeed! I realised it a bit after I wrote the comment. The problem is the code duplication between the "partial" and "full" expansion, both for `String_with_vars.t` and `Action_unexpanded.t.`

File: src/dune/build_system.mli (hunk 1 above)
Comment (diml) [reply src/dune/build_system.mli]: Indeed. That seems reasonable to me

**Comment by rgrinberg** [comment]
Long since done.

**Pull request closed without merging**